File: GuideObjC/GuideObjC/Objects/TableViewHandler.m
//
//  TableViewHandler.m
//  GuideObjC
//
//  Created by LuzanovRoman on 15.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "TableViewHandler.h"
#import <UIKit/UIKit.h>

@interface TableViewHandler ()
@property (nonatomic, retain) NSMutableArray * graphObjects;
@property (nonatomic, assign) NSInteger totalCount;
@property (nonatomic, weak) UITableView *tableView;
@end

@implementation TableViewHandler

- (id)init {
    self = [super init];
    
    if (self) {
        _graphObjects = [NSMutableArray array];
        _totalCount = -1;
    }
    return self;
}

- (id)initWithTableView:(UITableView *)tableView {
    self = [self init];
    self.tableView = tableView;
    return self;
}

- (NSObject *)objectAtIndex:(NSInteger)index {
    return _graphObjects[index];
}

- (NSObject *)last {
    return _graphObjects.lastObject;
}

- (NSInteger)count {
    return _graphObjects.count;
}

- (BOOL)isDownloaded {
    return _graphObjects.count == _totalCount;
}

- (void)setObjects:(NSArray *)objects totalCount:(NSInteger)count {
    if (objects) {
        [_graphObjects removeAllObjects];
        [_tableView reloadData];
        [self addObjects:objects totalCount:count];
    }
}

- (void)addObjects:(NSArray *)objects totalCount:(NSInteger)count {
    if (objects) {
        NSInteger start = _graphObjects.count;
        NSInteger end = start + objects.count;
        NSMutableArray *indexPaths = [NSMutableArray array];

        for (NSInteger i = start; i < end; i++) {
            [indexPaths addObject:[NSIndexPath indexPathForRow:i inSection:0]];
        }
        _totalCount = count;
        [_graphObjects addObjectsFromArray:objects];

        [_tableView beginUpdates];
        [_tableView insertRowsAtIndexPaths:indexPaths withRowAnimation:UITableViewRowAnimationNone];
        [_tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:0 inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
        [_tableView endUpdates];
    }
}

- (void)insertObject:(NSObject *)object atIndex:(NSInteger)index {
    if (object) {
        _totalCount += 1;
        [_graphObjects insertObject:object atIndex:index];
        
        [_tableView beginUpdates];
        [_tableView insertRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
        [_tableView endUpdates];
    }
}
@end

File: GuideSwift/GuideSwift/Objects/TableViewHandler.swift
//
//  TableViewHandler.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 15.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class TableViewHandler <T> {
    
    var graphObjects = [T]()
    var totalCount = -1
    weak var tableView: UITableView?
    
    init(withTableView tableView: UITableView) {
        self.tableView = tableView
    }
    
    subscript(index: Int) -> T {
        get {
            return graphObjects[index]
        }
    }
    
    var last: T? {
        get {
            return graphObjects.last
        }
    }
    
    var count: Int {
        get {
            return graphObjects.count
        }
    }
    
    var isDownloaded: Bool {
        get {
            return graphObjects.count == totalCount
        }
    }
    
    var noAnyData: Bool {
        get {
            return totalCount == -1
        }
    }
    
    func set(objects: [T]?, totalCount count: Int) {
        guard let theObjects = objects else { return }
        graphObjects.removeAll()
        tableView?.reloadData()
        add(objects: theObjects, totalCount: count)
    }
    
    func add(objects: [T]?, totalCount count: Int) {
        guard let theObjects = objects else { return }
        
        let start = graphObjects.count
        let end = start + theObjects.count
        var indexPaths = [IndexPath]()
        
        for i in start..<end {
            indexPaths.append(IndexPath(row: i, section: 0))
        }
        totalCount = count
        graphObjects.append(contentsOf: theObjects)
        
        tableView?.beginUpdates()
        tableView?.insertRows(at: indexPaths, with: .none)
        tableView?.reloadRows(at: [IndexPath(row: 0, section: 1)], with: .none)
        tableView?.endUpdates()
    }
    
    func insert(object: T?, at index:Int) {
        guard let theObject = object else { return }
        
        totalCount += 1
        graphObjects.insert(theObject, at: index)
        
        tableView?.beginUpdates()
        tableView?.insertRows(at: [IndexPath(row: index, section: 0)], with: .none)
        tableView?.endUpdates()
    }
}
